src/pybind/mgr/dashboard/settings.py

```diff
@@ -98,6 +98,10 @@ def __getattr__(cls, attr):
             value = False
         else:
             value = stype(mgr.get_module_option(attr, default))
+
+        if value == default and attr in ['GRAFANA_API_URL']:
+            value = stype(mgr.get_module_option_ex('orchestrator', attr, default))
```

Comment:
Please, don't do this:

* this whole `Settings` class should be deprecated and not extended (it made sense in the early days of dashboard, when `OPTIONS` types and defaults were not properly enforced by ceph-mgr C++ code, but since that changed this has caused a few issues as it doesn't match the behaviour of the C++ counterpart).

* this isn't solving the root issue (FQDN URLs), but just forcing users to manually hack it. If we are using an automated/smart deployment tool, I'd expect that this tool will at least take care for assigning usable URLs to services.

* it creates operational debt: users need to remember modifying almost the same setting in 2 different places (cephadm and dashboard). And, will this be the same for Prometheus/Alertmanager/...?

* it's hacky: breaks module isolation (what happens if the orchestrator module is not running?), modifies a behaviour that is not obvious neither for users nor dashboard devels, ...

Reply:
Big, thanks for the review!

(shuffled the order of bullet points)
> Please, don't do this:
> 
> * it creates operational debt: users need to remember modifying almost the same setting in 2 different places (cephadm and dashboard). And, will this be the same for Prometheus/Alertmanager/...?

The orchestrator setting is internal to the orchestrator and automatically set by cephadm. Users should not need to know its existence. I only care about the dashboard here.

The idea is to make the dashboard default to cephadm's url, without breaking any existing functionality.

> * this isn't solving the root issue (FQDN URLs), but just forcing users to manually hack it. If we are using an automated/smart deployment tool, I'd expect that this tool will at least take care for assigning usable URLs to services.

Hm, the user might **need** to access Grafana via the fqdn, while cephadm doesn't need to know the fqdn at all. Making the fqdn a requirement for cephadm just to make mgr/dashboard happy seems to be wrong to me. The goal of this PR is to no longer break the dashboard so that users still can overwrite things using the dashboard documentation.

Right now, we're breaking the dashboard by making this setting unusable. I don't want that.

> * it's hacky: breaks module isolation (what happens if the orchestrator module is not running?), modifies a behaviour that is not obvious neither for users nor dashboard devels, ...

Yeah, we need a simple API between mgr/dashboard and the orchestrator. I for one dislike the current approach of defining an api:

https://github.com/ceph/ceph/blob/36ddb9ed3b8a19ffcbd4b71817bc97df9cd07ac9/src/pybind/mgr/orchestrator/_interface.py#L1379-L1381

and #33947

As this requires the introduction of all possible values to `ServiceDescription`. 

> * this whole `Settings` class should be deprecated and not extended (it made sense in the early days of dashboard, when `OPTIONS` types and defaults were not properly enforced by ceph-mgr C++ code, but since that changed this has caused a few issues as it doesn't match the behaviour of the C++ counterpart).

Do you have a better idea?

Reply:
> Right now, we're breaking the dashboard by making this setting unusable. I don't want that.

Neither do I, but I dislike more the idea of creating a wormhole from the dashboard code into the orchestrator just to fix an integration issue.

Part of the disagreement here comes, and please correct me if I'm wrong, from the different views on what's the target deployment scenario:
* You maintain that the need of FQDN is exceptional and hence not a concern for cephadm,
* I think that FQDN should be the rule (versus short hostnames or IP addresses) and therefore a key requirement for cephadm. IMHO if you support FQDN, you can take the hostname/IP, but you cannot easily take the other way around.

Now focusing on the specific fix here: I think it hacks its own design. It mixes explicit and implicit configuration. Choose either: cephadm explicitly sets ceph-dashboard settings, or all dashboard settings are implicitly proxied to orchestrator mirror settings (and can also be manually overridden).

Reply:
> > Right now, we're breaking the dashboard by making this setting unusable. I don't want that.
> 
> Neither do I, but I dislike more the idea of creating a wormhole from the dashboard code into the orchestrator just to fix an integration issue.
> 
> Part of the disagreement here comes, and please correct me if I'm wrong, from the different views on what's the target deployment scenario:
> 
> * You maintain that the need of FQDN is exceptional and hence not a concern for cephadm,
> * I think that FQDN should be the rule (versus short hostnames or IP addresses) and therefore a key requirement for cephadm. IMHO if you support FQDN, you can take the hostname/IP, but you cannot easily take the other way around.

Hm, I was more looking into the general issue of not breaking the [dasbhaord UX](https://docs.ceph.com/docs/master/mgr/dashboard/#dashboard-grafana), than to find a fix for fqdns specifically. 

> 
> Now focusing on the specific fix here: I think it hacks its own design.

we need some kind of API here. cephadm can provide defaults for a lot of settings:

(no idea why the RGW url was broken apart)
* RGW_API_SCHEME
* RGW_API_HOST
* RGW_API_PORT
* RGW_API_ADMIN_RESOURCE

* RGW_API_ACCESS_KEY
* RGW_API_SECRET_KEY
* RGW_API_USER_ID
* RGW_API_SSL_VERIFY

* GRAFANA_API_URL
* GRAFANA_API_USERNAME
* GRAFANA_API_PASSWORD
* GRAFANA_API_SSL_VERIFY

* GANESHA_CLUSTERS_RADOS_POOL_NAMESPACE

* PROMETHEUS_API_HOST
* ALERTMANAGER_API_HOST

* ISCSI_API_SSL_VERIFICATION

(for haproxy:)
* config-key mgr/dashboard/ssl
* config-key mgr/dashboard/standby_behaviour

I need a way to provide default for all of them. not just GRAFANA_API_URL. 

> It mixes explicit and implicit configuration. Choose either: cephadm explicitly sets ceph-dashboard settings, or all dashboard settings are implicitly proxied to orchestrator mirror settings (and can also be manually overridden).

I'm against explicitly setting all of those mgr/dashboard settings, because then there is no way for users to modify them anymore. Then cephadm has to get everything right and 100% complete for each and every user out there.

Reply:
also, users should be able to use the dashboard settings to set things in the dashboard. Meaning, the dashboard documentation should continue to work.

Reply:
> I'm against explicitly setting all of those mgr/dashboard settings, because then there is no way for users to modify them anymore. Then cephadm has to get everything right and 100% complete for each and every user out there

> also, users should be able to use the dashboard settings to set things in the dashboard.

What I don't understand here is **why users would (imperatively) want to change a setting directly into the dashboard right after they used a declarative deployment tool** (let that be rook or cephadm). That sounds to me that the deployment tool has some work yet to do.

AFAICS cephadm is so far configuring things in the same way as rook does (and Ceph-Ansible, and DeepSea?): by invoking ceph commands and setting options. That's an explicit push-like API contract. That works for all the settings you enumerated there.

On the other hand, injecting the orchestrator as a man-in-the-middle between ceph-dashboard (or any other ceph-mgr module) and the persistence layer (the config store in this case) sounds really hacky to me. It mixes push and pull configuration approaches, which is weird.

More specific things:
* Many config settings need to trigger side-effects on changes, and they only happen when invoked from commands, so by patching the config store interface, in the best case, will changes go unnoticed to dashboard.
* Additionally, the `Settings` helper does only expose a few config settings, so you will need to globally "patch" `get_module_option()` and *conditionally* re-route all settings to either real dashboard settings or orchestrator clones.
* `get_module_option_ex` does not support localized settings, so you will to implement that new function as well, which will be interesting as it will need to check 4 different config options (a localized setting for dashboard, a general dashboard one, the same localized for orchestrator and finally the global orchestrator one).
* ceph has enough configuration layers (ceph.conf, daemon, config store, ...) so adding a new layer (orchestrator overrides) on top of those... is this really necessary?

Reply:
> What I don't understand here is **why users would (imperatively) want to change a setting directly into the dashboard right after they used a declarative deployment tool** (let that be rook or cephadm). That sounds to me that the deployment tool has some work yet to do.

I for one can't foresee all the different deployment setups out there. If you really want to break the UX by removing the possibility for users to change dashboard settings (cc @LenzGr ), we should probably restructure the dashboard documentation to make it clear they cannot change dashboard settings when running within a cephadm managed cluster. Wich might actually be a good thing. It's just bold. 

> AFAICS cephadm is so far configuring things in the same way as rook does (and Ceph-Ansible, and DeepSea?):
by invoking ceph commands and setting options. That's an explicit push-like API contract. That works for all the settings you enumerated there.

Sounds sensible. It's then up to the dasbhoard to sort out the fqdn issue. cephadm doesn't need fqdns and should not be forced to do so, just to let dashboard users access grafana.